File: src/main/java/data/model/SimpleTransforms.java
package data.model;

import data.Line;

import java.util.Random;

public class SimpleTransforms {
    public static void normalize(Line line, double scale){
        double minVal = line.getY(0);
        double maxVal = line.getY(0);
        final int size = line.getSize();
        for (int i = 1; i < size; i++) {
            double curr = line.getY(i);
            if (curr > maxVal) maxVal = curr;
            if (curr < minVal) minVal = curr;
        }
        for (int i = 0; i < size; i++) {
            double curr = line.getY(i);
            curr = scale*((curr-minVal)/(maxVal-minVal)-0.5);
            line.setY(i,curr);
        }
    }

    public static void spikes(Line line, int seed, int spikeNum, double scale){
        final Random rnd = new Random(seed);
        final int size = line.getSize();
        final double halfScale = scale/2.0;
        for (int i = spikeNum; i > 0; i--) {
            final int spikeIdx = rnd.nextInt(size);
            final double value = line.getY(spikeIdx);
            if (spikeIdx != 0) {
                final double y = line.getY(spikeIdx-1);
                line.setY(spikeIdx - 1,y+y*halfScale);
            }
            if (spikeIdx != size) {
                final double y = line.getY(spikeIdx+1);
                line.setY(spikeIdx + 1,y+y*halfScale);
            }
            line.setY(spikeIdx, value+value*scale);
        }
    }

    public static void shift(Line line, double start, double end, double shift, double scale){
        final int size = line.getSize();
        int i = 0;
        double curr = line.getX(i);
        while (i < size && curr < start){
            i++;
            curr = line.getX(i);
        }
        while (i < size && curr < end){
            final double val = line.getY(i);
            curr = line.getX(i);
            line.setY(i, (val+shift)*scale);
            i++;
        }
    }
}

File: src/main/java/data/model/SimpleTransforms.kt
package data.model

import data.Line

import java.util.Random

object SimpleTransforms {
    fun normalize(line: Line, scale: Double) {
        var minVal = line.getY(0)
        var maxVal = line.getY(0)
        val size = line.size
        for (i in 1 until size) {
            val curr = line.getY(i)
            if (curr > maxVal) maxVal = curr
            if (curr < minVal) minVal = curr
        }
        for (i in 0 until size) {
            var curr = line.getY(i)
            curr = scale * ((curr - minVal) / (maxVal - minVal) - 0.5)
            line.setY(i, curr)
        }
    }

    fun spikes(line: Line, seed: Int, spikeNum: Int, scale: Double) {
        val rnd = Random(seed.toLong())
        val size = line.size
        val halfScale = scale * 0.5
        for (i in spikeNum downTo 1) {
            val spikeIdx = rnd.nextInt(size)
            val value = line.getY(spikeIdx)
            if (spikeIdx != 0) {
                val y = line.getY(spikeIdx - 1)
                line.setY(spikeIdx - 1, y + y * halfScale)
            }
            if (spikeIdx != size) {
                val y = line.getY(spikeIdx + 1)
                line.setY(spikeIdx + 1, y + y * halfScale)
            }
            line.setY(spikeIdx, value + value * scale)
        }
    }

    fun shift(line: Line, start: Double, end: Double, shift: Double, scale: Double) {
        val size = line.size
        var i = 0
        var curr = line.getX(i)
        while (i < size && curr < start) {
            i++
            curr = line.getX(i)
        }
        while (i < size && curr < end) {
            val `val` = line.getY(i)
            curr = line.getX(i)
            line.setY(i, (`val` + shift) * scale)
            i++
        }
    }
}
